File: macos/Services/MCP/Security/MCPRateLimiter.swift
// MCPRateLimiter.swift
// Gridex
//
// Rate limiter for MCP operations.

import Foundation

actor MCPRateLimiter {
    private var queryCount: [UUID: RateBucket] = [:]
    private var writeCount: [UUID: RateBucket] = [:]
    private var ddlCount: [UUID: RateBucket] = [:]

    // Configurable limits (can be overridden via UserDefaults)
    var queriesPerMinute: Int { UserDefaults.standard.integer(forKey: "mcp.rateLimit.queriesPerMinute").nonZero ?? 60 }
    var queriesPerHour: Int { UserDefaults.standard.integer(forKey: "mcp.rateLimit.queriesPerHour").nonZero ?? 1000 }
    var writesPerMinute: Int { UserDefaults.standard.integer(forKey: "mcp.rateLimit.writesPerMinute").nonZero ?? 10 }
    var ddlPerMinute: Int { UserDefaults.standard.integer(forKey: "mcp.rateLimit.ddlPerMinute").nonZero ?? 1 }

    func checkLimit(tier: MCPPermissionTier, connectionId: UUID) -> RateLimitResult {
        let now = Date()

        switch tier {
        case .schema, .read, .advanced:
            return checkQueryLimit(connectionId: connectionId, now: now)
        case .write:
            return checkWriteLimit(connectionId: connectionId, now: now)
        case .ddl:
            return checkDDLLimit(connectionId: connectionId, now: now)
        }
    }

    func recordUsage(tier: MCPPermissionTier, connectionId: UUID) {
        let now = Date()

        switch tier {
        case .schema, .read, .advanced:
            recordQuery(connectionId: connectionId, now: now)
        case .write:
            recordWrite(connectionId: connectionId, now: now)
        case .ddl:
            recordDDL(connectionId: connectionId, now: now)
        }
    }

    // MARK: - Query Rate Limiting

    private func checkQueryLimit(connectionId: UUID, now: Date) -> RateLimitResult {
        let bucket = queryCount[connectionId] ?? RateBucket()

        // Check per-minute limit
        let minuteCount = bucket.countInWindow(now: now, windowSeconds: 60)
        if minuteCount >= queriesPerMinute {
            let retryAfter = bucket.oldestInWindow(now: now, windowSeconds: 60)
                .map { 60 - Int(now.timeIntervalSince($0)) } ?? 60
            return .exceeded(retryAfter: max(1, retryAfter))
        }

        // Check per-hour limit
        let hourCount = bucket.countInWindow(now: now, windowSeconds: 3600)
        if hourCount >= queriesPerHour {
            let retryAfter = bucket.oldestInWindow(now: now, windowSeconds: 3600)
                .map { 3600 - Int(now.timeIntervalSince($0)) } ?? 3600
            return .exceeded(retryAfter: max(1, retryAfter))
        }

        return .allowed
    }

    private func recordQuery(connectionId: UUID, now: Date) {
        var bucket = queryCount[connectionId] ?? RateBucket()
        bucket.record(now: now)
        bucket.cleanup(now: now, maxAge: 3600) // Keep 1 hour of history
        queryCount[connectionId] = bucket
    }

    // MARK: - Write Rate Limiting

    private func checkWriteLimit(connectionId: UUID, now: Date) -> RateLimitResult {
        let bucket = writeCount[connectionId] ?? RateBucket()

        let minuteCount = bucket.countInWindow(now: now, windowSeconds: 60)
        if minuteCount >= writesPerMinute {
            let retryAfter = bucket.oldestInWindow(now: now, windowSeconds: 60)
                .map { 60 - Int(now.timeIntervalSince($0)) } ?? 60
            return .exceeded(retryAfter: max(1, retryAfter))
        }

        return .allowed
    }

    private func recordWrite(connectionId: UUID, now: Date) {
        var bucket = writeCount[connectionId] ?? RateBucket()
        bucket.record(now: now)
        bucket.cleanup(now: now, maxAge: 60)
        writeCount[connectionId] = bucket
    }

    // MARK: - DDL Rate Limiting

    private func checkDDLLimit(connectionId: UUID, now: Date) -> RateLimitResult {
        let bucket = ddlCount[connectionId] ?? RateBucket()

        let minuteCount = bucket.countInWindow(now: now, windowSeconds: 60)
        if minuteCount >= ddlPerMinute {
            let retryAfter = bucket.oldestInWindow(now: now, windowSeconds: 60)
                .map { 60 - Int(now.timeIntervalSince($0)) } ?? 60
            return .exceeded(retryAfter: max(1, retryAfter))
        }

        return .allowed
    }

    private func recordDDL(connectionId: UUID, now: Date) {
        var bucket = ddlCount[connectionId] ?? RateBucket()
        bucket.record(now: now)
        bucket.cleanup(now: now, maxAge: 60)
        ddlCount[connectionId] = bucket
    }

    func resetLimits(for connectionId: UUID) {
        queryCount[connectionId] = nil
        writeCount[connectionId] = nil
        ddlCount[connectionId] = nil
    }

    func resetAllLimits() {
        queryCount.removeAll()
        writeCount.removeAll()
        ddlCount.removeAll()
    }
}

struct RateBucket {
    private var timestamps: [Date] = []

    mutating func record(now: Date) {
        timestamps.append(now)
    }

    func countInWindow(now: Date, windowSeconds: TimeInterval) -> Int {
        let cutoff = now.addingTimeInterval(-windowSeconds)
        return timestamps.filter { $0 > cutoff }.count
    }

    func oldestInWindow(now: Date, windowSeconds: TimeInterval) -> Date? {
        let cutoff = now.addingTimeInterval(-windowSeconds)
        return timestamps.filter { $0 > cutoff }.min()
    }

    mutating func cleanup(now: Date, maxAge: TimeInterval) {
        let cutoff = now.addingTimeInterval(-maxAge)
        timestamps = timestamps.filter { $0 > cutoff }
    }
}

enum RateLimitResult {
    case allowed
    case exceeded(retryAfter: Int)

    var isAllowed: Bool {
        if case .allowed = self { return true }
        return false
    }

    var retryAfterSeconds: Int? {
        if case .exceeded(let seconds) = self { return seconds }
        return nil
    }
}

private extension Int {
    var nonZero: Int? {
        self > 0 ? self : nil
    }
}

File: windows/Gridex/Services/MCP/Security/MCPRateLimiter.cpp
//
// MCPRateLimiter.cpp
//

#include "MCPRateLimiter.h"
#include <algorithm>

namespace DBModels
{
    using TimePoint = std::chrono::system_clock::time_point;

    // ── RateBucket ───────────────────────────────────────────

    int MCPRateLimiter::RateBucket::countInWindow(TimePoint now, int windowSeconds) const
    {
        const auto cutoff = now - std::chrono::seconds(windowSeconds);
        int n = 0;
        for (const auto& t : stamps) if (t > cutoff) ++n;
        return n;
    }

    TimePoint MCPRateLimiter::RateBucket::oldestInWindow(TimePoint now, int windowSeconds) const
    {
        const auto cutoff = now - std::chrono::seconds(windowSeconds);
        TimePoint best{};
        bool have = false;
        for (const auto& t : stamps)
        {
            if (t > cutoff && (!have || t < best)) { best = t; have = true; }
        }
        return have ? best : now;
    }

    void MCPRateLimiter::RateBucket::cleanup(TimePoint now, int maxAgeSeconds)
    {
        const auto cutoff = now - std::chrono::seconds(maxAgeSeconds);
        stamps.erase(
            std::remove_if(stamps.begin(), stamps.end(),
                [cutoff](const TimePoint& t) { return t <= cutoff; }),
            stamps.end());
    }

    // ── Public API ───────────────────────────────────────────

    void MCPRateLimiter::setConfig(const MCPRateLimiterConfig& c)
    {
        std::lock_guard<std::mutex> lk(mtx_);
        cfg_ = c;
    }

    MCPRateLimitResult MCPRateLimiter::checkLimit(
        MCPPermissionTier tier, const std::wstring& id)
    {
        std::lock_guard<std::mutex> lk(mtx_);
        const auto now = std::chrono::system_clock::now();
        switch (tier)
        {
            case MCPPermissionTier::Schema:
            case MCPPermissionTier::Read:
            case MCPPermissionTier::Advanced:
                return checkQuery(id, now);
            case MCPPermissionTier::Write:
                return checkWrite(id, now);
            case MCPPermissionTier::DDL:
                return checkDDL(id, now);
        }
        return { true, 0 };
    }

    void MCPRateLimiter::recordUsage(MCPPermissionTier tier, const std::wstring& id)
    {
        std::lock_guard<std::mutex> lk(mtx_);
        const auto now = std::chrono::system_clock::now();
        auto record = [now](auto& map, const std::wstring& key, int maxAge)
        {
            auto& bucket = map[key];
            bucket.record(now);
            bucket.cleanup(now, maxAge);
        };
        switch (tier)
        {
            case MCPPermissionTier::Schema:
            case MCPPermissionTier::Read:
            case MCPPermissionTier::Advanced:
                record(queryBuckets_, id, 3600); break;
            case MCPPermissionTier::Write:
                record(writeBuckets_, id, 60); break;
            case MCPPermissionTier::DDL:
                record(ddlBuckets_, id, 60); break;
        }
    }

    void MCPRateLimiter::resetLimits(const std::wstring& id)
    {
        std::lock_guard<std::mutex> lk(mtx_);
        queryBuckets_.erase(id);
        writeBuckets_.erase(id);
        ddlBuckets_.erase(id);
    }

    void MCPRateLimiter::resetAllLimits()
    {
        std::lock_guard<std::mutex> lk(mtx_);
        queryBuckets_.clear();
        writeBuckets_.clear();
        ddlBuckets_.clear();
    }

    // ── Internal: per-tier check ─────────────────────────────
    // (callers hold mtx_)

    MCPRateLimitResult MCPRateLimiter::checkQuery(const std::wstring& id, TimePoint now)
    {
        auto& b = queryBuckets_[id];
        const int minuteCount = b.countInWindow(now, 60);
        if (minuteCount >= cfg_.queriesPerMinute)
        {
            const auto oldest = b.oldestInWindow(now, 60);
            const int retry = 60 - static_cast<int>(std::chrono::duration_cast<std::chrono::seconds>(now - oldest).count());
            return { false, std::max(1, retry) };
        }
        const int hourCount = b.countInWindow(now, 3600);
        if (hourCount >= cfg_.queriesPerHour)
        {
            const auto oldest = b.oldestInWindow(now, 3600);
            const int retry = 3600 - static_cast<int>(std::chrono::duration_cast<std::chrono::seconds>(now - oldest).count());
            return { false, std::max(1, retry) };
        }
        return { true, 0 };
    }

    MCPRateLimitResult MCPRateLimiter::checkWrite(const std::wstring& id, TimePoint now)
    {
        auto& b = writeBuckets_[id];
        const int n = b.countInWindow(now, 60);
        if (n >= cfg_.writesPerMinute)
        {
            const auto oldest = b.oldestInWindow(now, 60);
            const int retry = 60 - static_cast<int>(std::chrono::duration_cast<std::chrono::seconds>(now - oldest).count());
            return { false, std::max(1, retry) };
        }
        return { true, 0 };
    }

    MCPRateLimitResult MCPRateLimiter::checkDDL(const std::wstring& id, TimePoint now)
    {
        auto& b = ddlBuckets_[id];
        const int n = b.countInWindow(now, 60);
        if (n >= cfg_.ddlPerMinute)
        {
            const auto oldest = b.oldestInWindow(now, 60);
            const int retry = 60 - static_cast<int>(std::chrono::duration_cast<std::chrono::seconds>(now - oldest).count());
            return { false, std::max(1, retry) };
        }
        return { true, 0 };
    }
}
